gateway/proxyproto/ssmproxy/ssm.go

```diff
@@ -0,0 +1,369 @@
+		grpc.WithOption(grpckey.ImpersonateUserSubjectHeaderKey, dbConnection.UserSubject),
+		grpc.WithOption("origin", pb.ConnectionOriginClient),
+		grpc.WithOption("verb", pb.ClientVerbConnect),
+		grpc.WithOption("session-id", sessionID),
+	)
+	if err != nil {
+		log.With("sid", sessionID, "conn", cID).Errorf("failed connecting to hoop server, reason=%v", err)
+		c.String(http.StatusInternalServerError, "Failed to connect to hoop server")
+	}
+	defer client.Close()
+
+	// Send an open session packet
+	err = client.Send(&pb.Packet{
+		Type: pbagent.SessionOpen,
+		Spec: map[string][]byte{
+			pb.SpecGatewaySessionID:   []byte(sessionID),
+			pb.SpecClientConnectionID: []byte(cID),
+		},
+	})
+	if err != nil {
+		log.With("sid", sessionID, "conn", cID).
+			Errorf("failed sending open session packet to hoop server, reason=%v", err)
+		return
+	}
+
+	// Wait for session open confirmation
+	pkt, err := client.Recv()
+	if err != nil {
+		log.With("sid", sessionID, "conn", cID).
+			Errorf("failed receiving session open confirmation from hoop server, reason=%v", err)
+		return
+	}
+
+	connectionType := pb.ConnectionType(pkt.Spec[pb.SpecConnectionType])
+	if pkt.Type != pbclient.SessionOpenOK || connectionType != pb.ConnectionTypeSSM {
+		log.With("sid", sessionID, "conn", cID).
+			Errorf("unsupported connection type, got=%v", connectionType)
+		return
+	}
+
+	log.With("sid", sessionID, "conn", cID).Debugf("Starting pipes")
+	err = sendWebsocketMessageHelper(client, msgType, msg, target, sessionID, cID)
+	if err != nil {
+		log.With("sid", sessionID, "conn", cID).
+			Errorf("error sending initial websocket message: %v", err)
+		return
+	}
+
+	// Ready for pumping
+	// Start RX Pipe (Client -> GRPC)
+	go r.handleRXPipe(ws, client, target, sessionID, cID)
+	// Start TX Pipe (GRPC -> Client)
+	r.handleTXPipe(ws, client, sessionID, cID)
+
+	log.With("sid", sessionID, "conn", cID).
+		Infof("connection closed for connectionId=%s, target=%s, sessionID=%s", connectionId, target, sessionID)
+}
+
+func (r *SSMProxy) handleTXPipe(ws *websocket.Conn, client pb.ClientTransport, sessionID, cID string) {
+	defer ws.Close()
+	defer client.Close()
+	// We need to read GRPC from another routine and pipe channel here
+	// Because otherwise we can't do websocket ping packets.
+	running := &atomic.Bool{}
+	running.Store(true)
+	packetChan := make(chan *pb.Packet)
+
+	const wsPingInterval = time.Second * 5
+
+	ticker := time.NewTicker(wsPingInterval)
+	defer ticker.Stop()
+
+	go func() {
+		for running.Load() {
+			msg, err := client.Recv()
+			if err != nil {
+				log.With("sid", sessionID, "conn", cID).Errorf("failed to receive packet from hoop server, reason=%v", err)
+				running.Store(false)
+			}
+			packetChan <- msg
+		}
+	}()
+
+	for running.Load() {
+		select {
+		case msg := <-packetChan:
+			switch msg.Type {
+			case pbclient.SSMConnectionWrite:
+				err := ws.WriteMessage(int(msg.Spec[pb.SpecWebsocketMessageType][0]), msg.Payload)
```

Comment:
Validate the length of `msg.Spec[pb.SpecWebsocketMessageType]` or it could panic!